### PR TITLE
[SYCL] Implement diagnostic for !access_mode::read + const T

### PR DESCRIPTION
**Pull request by mmoadeli**
According to [Read only accessors](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_read_only_accessors) A const qualified DataT is only allowed for a read-only accessor. 
This PR:
- Implements Implement diagnostic for !access_mode::read + const T
- Adds test to for above diagnostic
- Updates affected tests.
- Assigns access::mode:read to access mode of local accessor when the data type is constant. The previous value was access::mode:read, which lead to an assert after implementing above diagnostic.

### CODE REVIEW

File: sycl/test/basic_tests/accessor/const-type-non-readonly-accessor.cpp [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    defaultQueue.submit([&](handler &cgh) {
+      sycl::accessor accA{bufA, cgh, read_write};
+    });
+  } catch (const exception &e) {
+    std::cout << "Exception caught: " << e.what() << std::endl;
+  }
+  return 0;
+}
+
+// CHECK: static assertion failed due to requirement '!isConst || IsAccessReadOnly': A const qualified DataT is only allowed for a read-only accessor
```
[the remark this answers is not in the document]
Comment (mmoadeli) [reply sycl/test/basic_tests/accessor/const-type-non-readonly-accessor.cpp]: thanks @steffenlarsen , done

File: sycl/include/sycl/accessor.hpp [hunk 2]
```diff
@@ -2641,15 +2645,21 @@ class __SYCL_EBO __SYCL_SPECIAL_CLASS accessor<
 #endif
 };
 
+template <typename DataT> constexpr access::mode accessMode() {
```
[the remark this answers is not in the document]
Comment (mmoadeli) [reply sycl/include/sycl/accessor.hpp]: thanks @steffenlarsen , done

File: sycl/test/basic_tests/accessor/const-type-non-readonly-accessor.cpp [hunk 3]
```diff
@@ -1,5 +1,5 @@
-// RUN: not %clangxx -fsycl -fsycl-device-only -fsyntax-only \
-// RUN: %s -I %sycl_include 2>&1 | FileCheck %s
+// RUN: %clangxx -fsyntax-only %fsycl-host-only -Xclang -verify -Xclang -verify-ignore-unexpected=error,note %s
+// RUN: %clangxx -fsyntax-only -fsycl -fsycl-device-only -Xclang -verify -Xclang -verify-ignore-unexpected=error,note %s
```
[the remark this answers is not in the document]
Comment (mmoadeli) [reply sycl/test/basic_tests/accessor/const-type-non-readonly-accessor.cpp]: @steffenlarsen, I reverted this test to its original state. Using **-Xclang -verify** requires handling below 8 errors using **expected-error**

` File /home/mahmoud/codeplay/moadeli-llvm/release/bin/./../include/sycl/accessor.hpp Line 1008: 'AS' is a protected member of 'sycl::detail::accessor_common<const int, 1, sycl::access::mode::read_write, sycl::access::target::global_buffer, sycl::access::placeholder::false_t>'
  File /home/mahmoud/codeplay/moadeli-llvm/release/bin/./../include/sycl/accessor.hpp Line 1019: no member named 'AccessorSubscript' in 'sycl::detail::accessor_common<const int, 1, sycl::access::mode::read_write, sycl::access::target::global_buffer, sycl::access::placeholder::false_t>'
  File /home/mahmoud/codeplay/moadeli-llvm/release/bin/./../include/sycl/accessor.hpp Line 1016: 'IsPlaceH' is a protected member of 'sycl::detail::accessor_common<const int, 1, sycl::access::mode::read_write, sycl::access::target::global_buffer, sycl::access::placeholder::false_t>'
  File /home/mahmoud/codeplay/moadeli-llvm/release/bin/./../include/sycl/accessor.hpp Line 1015: 'IsHostBuf' is a protected member of 'sycl::detail::accessor_common<const int, 1, sycl::access::mode::read_write, sycl::access::target::global_buffer, sycl::access::placeholder::false_t>'
  File /home/mahmoud/codeplay/moadeli-llvm/release/bin/./../include/sycl/accessor.hpp Line 1014: 'IsGlobalBuf' is a protected member of 'sycl::detail::accessor_common<const int, 1, sycl::access::mode::read_write, sycl::access::target::global_buffer, sycl::access::placeholder::false_t>'
  File /home/mahmoud/codeplay/moadeli-llvm/release/bin/./../include/sycl/accessor.hpp Line 1013: 'IsConstantBuf' is a protected member of 'sycl::detail::accessor_common<const int, 1, sycl::access::mode::read_write, sycl::access::target::global_buffer, sycl::access::placeholder::false_t>'
  File /home/mahmoud/codeplay/moadeli-llvm/release/bin/./../include/sycl/accessor.hpp Line 1011: 'IsAccessAnyWrite' is a protected member of 'sycl::detail::accessor_common<const int, 1, sycl::access::mode::read_write, sycl::access::target::global_buffer, sycl::access::placeholder::false_t>'
  File /home/mahmoud/codeplay/moadeli-llvm/release/bin/./../include/sycl/accessor.hpp Line 1012: 'IsAccessReadOnly' is a protected member of 'sycl::detail::accessor_common<const int, 1, sycl::access::mode::read_write, sycl::access::target::global_buffer, sycl::access::placeholder::false_t>'`

[the remark this answers is not in the document]
Comment (steffenlarsen) [reply sycl/test/basic_tests/accessor/const-type-non-readonly-accessor.cpp]: Would it help to just run `%clangxx -fsycl -fsyntax-only -Xclang -verify %s -Xclang -verify-ignore-unexpected=note`?

[the remark this answers is not in the document]
Comment (mmoadeli) [reply sycl/test/basic_tests/accessor/const-type-non-readonly-accessor.cpp]: @steffenlarsen  the above ones are errors. There are another around 20 note ones.
I can handle it using below
`// RUN: not %clangxx -fsycl -Xclang -verify \
// RUN: %s -I %sycl_include 2>&1 | FileCheck %s
`
but that requires keeping the **// CHECK**

File: sycl/include/sycl/accessor.hpp [hunk 4]
```diff
@@ -2643,13 +2666,16 @@ class __SYCL_EBO __SYCL_SPECIAL_CLASS accessor<
...
-    : public local_accessor_base<DataT, Dimensions, access::mode::read_write,
+    : public local_accessor_base<DataT, Dimensions, detail::accessModeFromConstness<DataT>(),
                                  access::placeholder::false_t>,
       public detail::OwnerLessBase<local_accessor<DataT, Dimensions>> {
 
-  using local_acc =
-      local_accessor_base<DataT, Dimensions, access::mode::read_write,
-                          access::placeholder::false_t>;
+  using local_acc = local_accessor_base<DataT, Dimensions, detail::accessModeFromConstness<DataT>(),
+                                        access::placeholder::false_t>;
```
Comment (steffenlarsen) [inline sycl/include/sycl/accessor.hpp:2674]: The formatting here looks off. Don't we have >80 character lines in the above changes? Weird that lint doesn't pick that up.

**Review by steffenlarsen — APPROVED** [review]
Apart from a small comment about formatting, my concerns have been addressed! 😄

File: sycl/test/basic_tests/accessor/const-type-non-readonly-accessor.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (steffenlarsen) [reply sycl/test/basic_tests/accessor/const-type-non-readonly-accessor.cpp]: Discussed offline. Issue found and eliminated.

File: sycl/include/sycl/accessor.hpp (hunk 4 above)
Comment (mmoadeli) [reply sycl/include/sycl/accessor.hpp:2674]: I always use git clang-format.  I'll fix it.

Comment (mmoadeli) [reply sycl/include/sycl/accessor.hpp]: thanks @steffenlarsen,   done

**Comment by bader** [comment]
Failed Tests (1):
  SYCL :: KernelAndProgram/spec_constants_after_link.cpp

**Pull request merged**